File: 12/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

const mod = 998244353

func main() {
	reader := bufio.NewReaderSize(os.Stdin, 1<<20)
	writer := bufio.NewWriterSize(os.Stdout, 1<<20)
	defer writer.Flush()

	// Читаем n и m
	line, _ := reader.ReadString('\n')
	parts := strings.Fields(strings.TrimSpace(line))
	n, _ := strconv.ParseInt(parts[0], 10, 64)
	m, _ := strconv.Atoi(parts[1])

	// Читаем хорошие числа
	line, _ = reader.ReadString('\n')
	parts = strings.Fields(strings.TrimSpace(line))
	good := make(map[int]bool, m)
	for i := 0; i < m; i++ {
		val, _ := strconv.Atoi(parts[i])
		good[val] = true
	}

	result := solve(int64(n), good)
	writer.WriteString(fmt.Sprintf("%d\n", result))
}

// solve находит количество чудесных чисел длины n
// Чудесное число: без лидирующих нулей, сумма любых трех последовательных цифр - хорошее число
func solve(n int64, good map[int]bool) int {
	if n < 3 {
		return 0
	}

	// Состояние: последние две цифры (d1, d2) -> индекс = d1*10 + d2
	// Матрица переходов: M[i][j] = 1, если можно перейти от состояния i к состоянию j
	// i = d1*10 + d2, j = d2*10 + d3, переход возможен если d1+d2+d3 - хорошее число

	// Строим матрицу переходов 100x100
	M := make([][]int, 100)
	for i := 0; i < 100; i++ {
		M[i] = make([]int, 100)
	}

	for d1 := 0; d1 < 10; d1++ {
		for d2 := 0; d2 < 10; d2++ {
			from := d1*10 + d2
			for d3 := 0; d3 < 10; d3++ {
				sum := d1 + d2 + d3
				if good[sum] {
					to := d2*10 + d3
					M[from][to] = 1
				}
			}
		}
	}

	// Начальный вектор: для чисел длины 2 (d1, d2), где d1 != 0
	start := make([]int, 100)
	for d1 := 1; d1 < 10; d1++ {
		for d2 := 0; d2 < 10; d2++ {
			idx := d1*10 + d2
			start[idx] = 1
		}
	}

	// Если n == 2, возвращаем количество начальных состояний
	if n == 2 {
		result := 0
		for i := 0; i < 100; i++ {
			result = (result + start[i]) % mod
		}
		return result
	}

	// Возводим матрицу в степень (n-2), так как у нас уже есть первые 2 цифры
	power := n - 2
	M_power := matrixPower(M, power)

	// Умножаем начальный вектор на матрицу
	result := 0
	for i := 0; i < 100; i++ {
		for j := 0; j < 100; j++ {
			result = (result + start[i]*M_power[i][j]%mod) % mod
		}
	}

	return result
}

// matrixPower возводит матрицу в степень используя быстрое возведение
func matrixPower(M [][]int, power int64) [][]int {
	n := len(M)

	// Инициализируем единичную матрицу
	result := make([][]int, n)
	for i := 0; i < n; i++ {
		result[i] = make([]int, n)
		result[i][i] = 1
	}

	base := make([][]int, n)
	for i := 0; i < n; i++ {
		base[i] = make([]int, n)
		copy(base[i], M[i])
	}

	// Быстрое возведение в степень
	for power > 0 {
		if power&1 == 1 {
			result = matrixMultiply(result, base)
		}
		base = matrixMultiply(base, base)
		power >>= 1
	}

	return result
}

// matrixMultiply умножает две матрицы по модулю
func matrixMultiply(A, B [][]int) [][]int {
	n := len(A)
	m := len(B[0])
	k := len(B)

	result := make([][]int, n)
	for i := 0; i < n; i++ {
		result[i] = make([]int, m)
		for j := 0; j < m; j++ {
			sum := 0
			for t := 0; t < k; t++ {
				sum = (sum + A[i][t]*B[t][j]%mod) % mod
			}
			result[i][j] = sum
		}
	}

	return result
}

File: 12/main.rs
use std::io::{self, BufRead, BufWriter, Write};

const MOD: i64 = 998244353;

fn main() {
    let stdin = io::stdin();
    let mut reader = stdin.lock();
    let stdout = io::stdout();
    let mut writer = BufWriter::new(stdout.lock());

    // Читаем n и m
    let mut line = String::new();
    reader.read_line(&mut line).unwrap();
    let parts: Vec<&str> = line.trim().split_whitespace().collect();
    if parts.len() < 2 {
        return;
    }
    let n: i64 = parts[0].parse().unwrap();
    let m: usize = parts[1].parse().unwrap();

    // Читаем хорошие числа
    let mut line = String::new();
    reader.read_line(&mut line).unwrap();
    let parts: Vec<&str> = line.trim().split_whitespace().collect();
    let mut good = std::collections::HashSet::new();
    for i in 0..m {
        if i < parts.len() {
            let val: i32 = parts[i].parse().unwrap();
            good.insert(val);
        }
    }

    let result = solve(n, &good);
    writeln!(writer, "{}", result).unwrap();
    writer.flush().unwrap();
}

// solve находит количество чудесных чисел длины n
// Чудесное число: без лидирующих нулей, сумма любых трех последовательных цифр - хорошее число
fn solve(n: i64, good: &std::collections::HashSet<i32>) -> i64 {
    if n < 3 {
        return 0;
    }

    // Состояние: последние две цифры (d1, d2) -> индекс = d1*10 + d2
    // Матрица переходов: M[i][j] = 1, если можно перейти от состояния i к состоянию j
    // i = d1*10 + d2, j = d2*10 + d3, переход возможен если d1+d2+d3 - хорошее число

    // Строим матрицу переходов 100x100
    let mut m = vec![vec![0i64; 100]; 100];

    for d1 in 0..10 {
        for d2 in 0..10 {
            let from = (d1 * 10 + d2) as usize;
            for d3 in 0..10 {
                let sum = d1 + d2 + d3;
                if good.contains(&sum) {
                    let to = (d2 * 10 + d3) as usize;
                    m[from][to] = 1;
                }
            }
        }
    }

    // Начальный вектор: для чисел длины 2 (d1, d2), где d1 != 0
    let mut start = vec![0i64; 100];
    for d1 in 1..10 {
        for d2 in 0..10 {
            let idx = (d1 * 10 + d2) as usize;
            start[idx] = 1;
        }
    }

    // Если n == 2, возвращаем количество начальных состояний
    if n == 2 {
        return start.iter().sum::<i64>() % MOD;
    }

    // Возводим матрицу в степень (n-2), так как у нас уже есть первые 2 цифры
    let power = n - 2;
    let m_power = matrix_power(&m, power);

    // Умножаем начальный вектор на матрицу
    let mut result = 0i64;
    for i in 0..100 {
        for j in 0..100 {
            result = (result + start[i] * m_power[i][j] % MOD) % MOD;
        }
    }

    result
}

// matrix_power возводит матрицу в степень используя быстрое возведение
fn matrix_power(m: &[Vec<i64>], power: i64) -> Vec<Vec<i64>> {
    let n = m.len();

    // Инициализируем единичную матрицу
    let mut result = vec![vec![0i64; n]; n];
    for i in 0..n {
        result[i][i] = 1;
    }

    let mut base = m.to_vec();
    let mut p = power;

    // Быстрое возведение в степень
    while p > 0 {
        if p & 1 == 1 {
            result = matrix_multiply(&result, &base);
        }
        base = matrix_multiply(&base, &base);
        p >>= 1;
    }

    result
}

// matrix_multiply умножает две матрицы по модулю
fn matrix_multiply(a: &[Vec<i64>], b: &[Vec<i64>]) -> Vec<Vec<i64>> {
    let n = a.len();
    let m = b[0].len();
    let k = b.len();

    let mut result = vec![vec![0i64; m]; n];

    for i in 0..n {
        for j in 0..m {
            let mut sum = 0i64;
            for t in 0..k {
                sum = (sum + a[i][t] * b[t][j] % MOD) % MOD;
            }
            result[i][j] = sum;
        }
    }

    result
}

